common/views/components/ImageViewer/ImageViewer.js

```diff
@@ -1,60 +1,148 @@
+      profile: ['http://iiif.io/api/image/2/level2.json'],
+      protocol: 'http://iiif.io/api/image',
+      tiles: [
+        {
+          scaleFactors: [1, 2, 4, 8, 16, 32],
+          width: 400,
+        },
+      ],
+    },
+  ];
+}
 
-type ViewerContentProps = {|
+type ImageViewerProps = {|
   id: string,
-  classes: string,
-  viewerVisible: boolean,
+  src: string,
+  srcSet: string,
+  width: number,
+  height?: number,
+  canvasOcr: ?string,
   infoUrl: string,
+  lang: ?string,
 |};
 
-const ViewerContent = ({
+const ImageViewer = ({
   id,
-  classes,
-  viewerVisible,
+  width,
+  height,
+  canvasOcr,
+  lang,
   infoUrl,
-}: ViewerContentProps) => {
-  const handleRotate = event => {
-    trackEvent({
-      category: 'Control',
-      action: 'rotate ImageViewer',
-      label: id,
-    });
-  };
-
-  const handleZoomIn = event => {
-    trackEvent({
-      category: 'Control',
-      action: 'zoom in ImageViewer',
-      label: id,
-    });
-  };
-
-  const handleZoomOut = event => {
-    trackEvent({
-      category: 'Control',
-      action: 'zoom out ImageViewer',
-      label: id,
-    });
-  };
+  src,
+  srcSet,
+}: ImageViewerProps) => {
+  const [viewer, setViewer] = useState(null);
+  const zoomStep = 0.5;
+
+  useEffect(() => {
+    if (viewer) {
+      fetch(infoUrl)
+        .then(response => response.json())
+        .then(data => {
+          viewer.open(getTileSources(data));
+        });
+    }
+  }, [infoUrl]);
+
+  async function setupViewer(imageInfoSrc, viewerId) {
+    const { default: OpenSeadragon } = await import('openseadragon');
+
+    return fetch(imageInfoSrc)
+      .then(response => response.json())
+      .then(data => {
+        const osdViewer = OpenSeadragon({
+          id: `image-viewer-${viewerId}`,
+          showNavigationControl: false,
+          gestureSettingsMouse: {
+            scrollToZoom: false,
+          },
+          tileSources: getTileSources(data),
+        });
+
+        setViewer(osdViewer);
+
+        return osdViewer;
+      })
+      .catch(error => {
+        Raven.captureException(new Error(`OpenSeadragon error: ${error}`), {
+          tags: {
+            service: 'dlcs',
+          },
+        });
+      });
+  }
+
+  async function handleRotate() {
+    const v = viewer || (await setupViewer(infoUrl, id));
+
+    v.viewport.setRotation(v.viewport.getRotation() + 90);
+  }
+
+  function doZoomIn(viewer) {
+    const max = viewer.viewport.getMaxZoom();
+    const nextMax = viewer.viewport.getZoom() + zoomStep;
+    const newMax = nextMax <= max ? nextMax : max;
+
+    viewer.viewport.zoomTo(newMax);
+  }
+
+  function doZoomOut(viewer) {
+    const min = viewer.viewport.getMinZoom();
+    const nextMin = viewer.viewport.getZoom() - zoomStep;
+    const newMin = nextMin >= min ? nextMin : min;
+
+    viewer.viewport.zoomTo(newMin);
+  }
+
+  async function handleZoomIn() {
+    const v = viewer || (await setupViewer(infoUrl, id));
+
+    if (v.isOpen()) {
```

Comment:
Could you run me through what this is doing?

Reply:
if the OSD viewer is open, we're ready to zoom straight away, so we pass the viewer to `doZoomIn` which does the zooming.

Otherwise, we wait for the `tile-loaded` event which signifies that the tiles are in memory allowing for us to zoom with animation.

If instead we just wait for the `open` event, we can zoom to the correct level, but because this will fire before the tiles are loaded, the first zoom will not have any animation.